crypto/rand/drbg_lib.c

```diff
@@ -141,6 +147,161 @@ static int is_digest(int type)
     }
 }
 
+/*
+ * Initialize the OPENSSL_CTX global DRBGs on first use.
+ * Returns the allocated global data on success or NULL on failure.
+ */
+static void *drbg_ossl_ctx_new(OPENSSL_CTX *libctx)
+{
+    DRBG_GLOBAL *dgbl = OPENSSL_zalloc(sizeof(*dgbl));
+
+    if (dgbl == NULL)
+        return NULL;
+
+#ifndef FIPS_MODE
+    /*
+     * We need to ensure that base libcrypto thread handling has been
+     * initialised.
+     */
+     OPENSSL_init_crypto(0, NULL);
+#endif
+
+    if (!CRYPTO_THREAD_init_local(&dgbl->private_drbg, NULL))
+        goto err1;
```

Comment:
I assume all these goto's mean that CRYPTO_THREAD_cleanup_local() doesnt handle passing NULLS - looks  messy

Reply:
Well, `dgbl->private_drbg` is not NULL if there's an error, just uninitialised - so we shouldn't attempt to use it in anyway.

Reply:
Since dgbl is allocated via zalloc, and since NULL is all-bits-zero (via the sanity test), it is initialized.

Reply:
It's the `private_drbg` field we are specifically concerned with there. And yes it will get set to 0 on create. However, when I said it was unitialised, I meant it has not been initialised via a call to `CRYPTO_THREAD_init_local`. We should not be calling `CRYPTO_THREAD_cleanup_local` on keys that have not been initialised in that way. We cannot know how `CRYPTO_THREAD_cleanup_local` would treat a key that is set to 0. It most likely would be fine but we cannot say for sure. For example in a pthreads implementation `CRYPTO_THREAD_cleanup_local` calls `pthread_key_delete` which has this description: "The pthread_key_delete() function shall delete a thread-specific data key previously returned by pthread_key_create()". It says nothing about what happens if you pass it zero data that was not created by `pthread_key_create` - so at best this behaviour is undefined.